runtime-modules/storage/src/lib.rs

```diff
@@ -3729,25 +3735,32 @@ impl<T: Trait> Module<T> {
 
     // Returns only existing hashes in the blacklist from the original collection.
     #[allow(clippy::redundant_closure)] // doesn't work with Substrate storage functions.
-    fn get_existing_hashes(hashes: &BTreeSet<Cid>) -> BTreeSet<Cid> {
+    fn get_existing_hashes(hashes: &BTreeSet<Cid>) -> Result<BTreeSet<Cid>, DispatchError> {
         Self::get_hashes_by_predicate(hashes, |cid| Blacklist::contains_key(cid))
     }
 
     // Returns only nonexisting hashes in the blacklist from the original collection.
-    fn get_nonexisting_hashes(hashes: &BTreeSet<Cid>) -> BTreeSet<Cid> {
+    fn get_nonexisting_hashes(hashes: &BTreeSet<Cid>) -> Result<BTreeSet<Cid>, DispatchError> {
         Self::get_hashes_by_predicate(hashes, |cid| !Blacklist::contains_key(cid))
     }
 
     // Returns hashes from the original collection selected by predicate.
     fn get_hashes_by_predicate<P: FnMut(&&Cid) -> bool>(
         hashes: &BTreeSet<Cid>,
         predicate: P,
-    ) -> BTreeSet<Cid> {
-        hashes
+    ) -> Result<BTreeSet<Cid>, DispatchError> {
+        ensure!(
+            hashes.iter().all(|cid| cid.len() == CID_LENGTH),
+            Error::<T>::InvalidCidLength
+        );
```

Comment:
Shouldn't we use a fixed-length type (like `Hash` or `[u8; 32]`) instead of checking if the provided vector is of expected length?

Reply:
`Vec` is a generally easier type to work with. We accept only `Vec` in all the runtime and I'm not sure we need to create an exception. Also, we have `Vec` also on the TypeScript side - we probably need to test the apps if we change the parameter type. I'd like to avoid it. If we change the [hashing algorithm](https://github.com/multiformats/multihash#example) it would be easier to relax the check to  the `less than` operator and support two hashing types.

Reply:
Ok, this change would probably have big implications across the entire project, so let's leave it as is.